docs/pages/indexing.md

```diff
@@ -1,18 +1,325 @@
+when indexing. For example if you wanted to get the first two columns, you would
+access the array by passing span for the first argument and seq(2) as the second
+argument.
+
+\snippet test/index.cpp index_tutorial_first_two_columns
+
+\f[
+A(span, seq(2)) =
+\begin{bmatrix}
+     0 & 4 \\
+     1 & 5 \\
+     2 & 6 \\
+     3 & 7
+\end{bmatrix}
+\f]
+
+The af::seq object can also be used to define ranges that are not continuous. There are
+three constructors for af::seq.
+
+* af::seq(N): Defines a range between 0 and N-1
+* af::seq(begin, end) Defines a range between begin and end inclusive
+* af::seq(begin, end, step) defines a range between begin and end striding by step values
+
+You can select the second and forth(last) rows by passing (seq(1, end, 2), span)
+to the indexing operator.
+
+\snippet test/index.cpp index_tutorial_second_and_fourth_rows
+
+\f[
+A(seq(1, end, 2), span) =
+\begin{bmatrix}
+     1 & 5 &  9 & 13 \\
+     3 & 7 & 11 & 15
+\end{bmatrix}
+\f]
+
+## Indexing using af::array
+
+You can also index using other af::array objects. ArrayFire performs a Cartesian
+product of the input arrays.
+
+\snippet test/index.cpp index_tutorial_array_indexing
+
+\f[
+A =
+\begin{bmatrix}
+    0 & 4 & 8 & 12 \\
+    1 & 5 & 9 & 13 \\
+    2 & 6 & 10 & 14 \\
+    3 & 7 & 11 & 15
+\end{bmatrix}
+\\
+A(
+\begin{bmatrix}
+2 \\ 1 \\ 3
+\end{bmatrix}
+,
+\begin{bmatrix}
+3 \\ 1 \\ 2
+\end{bmatrix}
+) =
+
+\begin{bmatrix}
+(2,3) & (2,1) & (2,2) \\
+(1,3) & (1,1) & (1,2) \\
+(3,3) & (3,1) & (3,2)
+\end{bmatrix}
+=
+\begin{bmatrix}
+14 & 6 & 10 \\
+13 & 5 &  9 \\
+15 & 7 & 11
+\end{bmatrix}
+\f]
+
+
+If you want to index an af::array using coordinate arrays, you can do that using the
+af::approx1 and af::approx2 functions.
+
+\snippet test/index.cpp index_tutorial_approx
+
+\f[
+approx2(A,
+\begin{bmatrix}
+2 \\ 1 \\ 3
+\end{bmatrix}
+,
+\begin{bmatrix}
+3 \\ 1 \\ 2
+\end{bmatrix}
+) =
+\begin{bmatrix}
+(2,3) \\
+(1,1) \\
+(3,2)
+\end{bmatrix}
+=
+\begin{bmatrix}
+14 \\
+ 5 \\
+11
+\end{bmatrix}
+\f]
+
+Boolean(b8) arrays can be used to index into another array. In this type of
+indexing the non-zero values will be selected by the boolean operation. If we
+want to select all values less than 5, we can pass a boolean expression into
+the parenthesis operator.
+
+\snippet test/index.cpp index_tutorial_boolean
+
+\f[
+out =
+\begin{bmatrix}
+0 \\ 1 \\ 2 \\ 3 \\ 4
+\end{bmatrix}
+\f]
+
+## References and copies
+
+All ArrayFire indexing functions return af::array(technically its an array_proxy
+class but we will discuss that later) objects. These objects may be new arrays
+or they may reference the original array depending on the type of indexing that
+was performed on them. If an array was indexed using another af::array or it
+was indexed using the af::approx functions, then a new array is created. It does
+not reference the original data.
+
+If an array was indexed using a scalar, af::seq or af::span, then the resulting
```

Comment:
This new paragraph is not necessary as we still continuing topic from first paragraph. We should either merge this into above paragraph or list the two types of outcome in a bullet list fashion after we end the sentence `that was performed on them.`.

Reply:
```
All ArrayFire indexing functions return af::array(technically its an array_proxy
class) objects. These objects may be new arrays or they may reference the
original array depending on the type of indexing that was performed on them. 

- If an array was indexed using another af::array or it was indexed using the
af::approx functions, then a new array is created. It does not reference the
original data.
- If an array was indexed using a scalar, af::seq or af::span, then
the resulting array will reference the original data IF the first dimension is
continuous. The following lines will not allocate additional memory.

\note The new arrays wither references or newly allocated arrays, are
independent of the original data. Meaning that any changes to the original array
will not propagate to the references. Likewise, any changes to the reference
arrays will not modify the original data.

\snippet test/index.cpp index_tutorial_references

The following code snippet shows some examples of indexing that will allocate
new memory.

\snippet test/index.cpp index_tutorial_copies

Notice that even though the copy3 array is referencing continuous memory in the
original array, a new array is created because we used an array to index into
the af::array.
```